utils/stall_free.hh

```diff
@@ -56,14 +58,93 @@ void merge_to_gently(std::list<T>& list1, const std::list<T>& list2, Compare com
     }
 }
 
-template<class T>
-seastar::future<> clear_gently(std::list<T>& list) {
-    return repeat([&list] () mutable {
-        if (list.empty()) {
-            return seastar::stop_iteration::yes;
-        }
-        list.pop_back();
-        return seastar::stop_iteration::no;
+template <typename T>
+concept GentlyClearable = requires (T x) {
+    { x.clear_gently() } -> std::same_as<future<>>;
+};
+
+template <typename T>
+concept SmartPointer = requires (T x) {
+    { x.reset() } -> std::same_as<void>;
+    { x.get() } -> std::same_as<typename T::element_type*>;
+};
+
+template <typename T>
+concept VectorLike = requires (T x, size_t n) {
+    { x.empty() } -> std::same_as<bool>;
+    { x.back() } -> std::same_as<typename T::value_type&>;
+    { x.pop_back() } -> std::same_as<void>;
+};
+
+template <typename T>
+concept Container = !VectorLike<T> && requires (T x) {
+    { x.empty() } -> std::same_as<bool>;
+    { x.erase(x.begin()) } -> std::same_as<typename T::iterator>;
+};
```

Comment:
`!VectorLike` constraint here is just for disambiguation since a `std::vector`, for example, matches both `VectorLike` and `Container` and the compiler complains about it.

`ListLike` is fine too.

Reply:
Certainly saying that a Container is !VectorLike is wrong.

Reply:
Move the disambiguation to the code that needs it, don't make the concept wrong.

Reply:
Fixed in v6